nussl/separation/duet.py

```diff
@@ -297,6 +356,63 @@ def find_peaks(data, min_thr=0.5, min_dist=None, max_num=1):
 
         return peak_indices
 
+    def compute_masks(self, atn_peak, delay_peak):
+        """Receives the attenuation and delay peaks and computes a mask to be applied to the signal for source
+        separation.
+
+        Parameters:
+            atn_peak (np.array): Attenuation peaks determined from histogram
+            delay_peak (np.array): Delay peaks determined from histogram
+        Returns:
+           best_ind(np.array): The indices for where the masks are applied
+           mask (np.array): A binary mask used to separate the sources
+
+        """
+        # compute masks for separation
+        best_so_far = np.inf * np.ones((self.num_frequency_bins - 1, self.num_time_bins))
+        best_ind = np.zeros((self.num_frequency_bins - 1, self.num_time_bins), dtype=int)
+        mask = None
+
+        for i in range(0, self.num_sources):
+            phase = np.exp(-1j * self.frequency_matrix * delay_peak[i])
+            score = np.abs(atn_peak[i] * phase * self.stft_ch0 - self.stft_ch1) ** 2 / (1 + atn_peak[i] ** 2)
+            mask = (score < best_so_far)
+            best_ind[mask] = i
+            best_so_far[mask] = score[mask]
+
+        return best_ind, mask
+
+    def convert_time_domain(self, best_ind, mask, atn_peak, delay_peak):
+        """Receives the attenuation and delay peaks, the mask and best indices and 
+        applies the mask to separate the sources
+
+        Parameters:
+            best_ind (np.array): The indices for where the masks are applied
+            mask (np.array): binary mask used to separate the sources
+            atn_peak (np.array): Attenuation peaks determined from histogram
+            delay_peak (np.array): Delay peaks determined from histogram
+
+        Returns:
+           source_estimates (np.array): The arrays of the separated source signals
+
+        """
+        source_estimates = np.zeros((self.num_sources, self.audio_signal.signal_length))
+        for i in range(0, self.num_sources):
+            mask = (best_ind == i)
+            Xm = np.vstack([np.zeros((1, self.num_time_bins)),
```

Comment:
`Xm` is a bad variable name. Also, get rid of this vstack (looping through a list might help here).